File: javascript/go-counting/go-counting.js
//
// This is only a SKELETON file for the 'Go Counting' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class GoCounting {
  constructor(board) {
    this.board = board.map(row => row.split(''));
    this.height = this.board.length;
    this.width = this.board[0].length;
  }

  isValidCoord(x, y) {
    return x >= 0 && x < this.width && y >= 0 && y < this.height;
  }

  getTerritory(x, y) {
    if (!this.isValidCoord(x, y)) {
      return { error: 'Invalid coordinate' };
    }

    const point = this.board[y][x];
    if (point === 'B' || point === 'W') {
      return { owner: 'NONE', territory: [] };
    }

    const visited = new Set();
    const territory = [];
    const borderColors = new Set();

    const stack = [[x, y]];
    while (stack.length > 0) {
      const [cx, cy] = stack.pop();
      const key = `${cx},${cy}`;
      if (visited.has(key)) continue;
      visited.add(key);

      const cell = this.board[cy][cx];
      if (cell === ' ') {
        territory.push([cx, cy]);
        // Explore neighbors
        const neighbors = [
          [cx - 1, cy],
          [cx + 1, cy],
          [cx, cy - 1],
          [cx, cy + 1],
        ];
        for (const [nx, ny] of neighbors) {
          if (!this.isValidCoord(nx, ny)) continue;
          const neighborCell = this.board[ny][nx];
          if (neighborCell === ' ') {
            stack.push([nx, ny]);
          } else if (neighborCell === 'B' || neighborCell === 'W') {
            borderColors.add(neighborCell);
          }
        }
      }
    }

    let owner = 'NONE';
    if (borderColors.size === 1) {
      const color = borderColors.values().next().value;
      owner = color === 'B' ? 'BLACK' : 'WHITE';
    }

    territory.sort((a, b) => {
      if (a[0] !== b[0]) {
        return a[0] - b[0]; // sort by x
      }
      return a[1] - b[1]; // then by y
    });

    return { owner, territory };
  }

  getTerritories() {
    const visited = new Set();
    const territoryBlack = [];
    const territoryWhite = [];
    const territoryNone = [];

    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const key = `${x},${y}`;
        if (visited.has(key)) continue;
        if (this.board[y][x] !== ' ') continue;

        const result = this.getTerritory(x, y);
        for (const [tx, ty] of result.territory) {
          visited.add(`${tx},${ty}`);
        }

        if (result.owner === 'BLACK') {
          territoryBlack.push(...result.territory);
        } else if (result.owner === 'WHITE') {
          territoryWhite.push(...result.territory);
        } else {
          territoryNone.push(...result.territory);
        }
      }
    }

    return {
      territoryBlack,
      territoryWhite,
      territoryNone,
    };
  }
}

File: java/go-counting/src/main/java/GoCounting.java
import java.awt.Point;
import java.util.Map;
import java.util.Set;

class GoCounting {

    private final char[][] grid;
    private final int rows;
    private final int cols;

    GoCounting(String board) {
        String[] lines = board.split("\n");
        rows = lines.length;
        cols = lines[0].length();
        grid = new char[rows][cols];
        for (int y = 0; y < rows; y++) {
            grid[y] = lines[y].toCharArray();
        }
    }

    private void validateCoordinate(int x, int y) {
        if (x < 0 || y < 0 || y >= rows || x >= cols) {
            throw new IllegalArgumentException("Invalid coordinate");
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (char[] row : grid) {
            sb.append(new String(row)).append("\n");
        }
        return sb.toString();
    }

    Player getTerritoryOwner(int x, int y) {
        validateCoordinate(x, y);
        if (grid[y][x] == 'B' || grid[y][x] == 'W') {
            return Player.NONE;
        }
        TerritoryResult result = floodFill(x, y, new boolean[rows][cols]);
        return result.owner;
    }

    Set<Point> getTerritory(int x, int y) {
        validateCoordinate(x, y);
        if (grid[y][x] == 'B' || grid[y][x] == 'W') {
            return Set.of();
        }
        TerritoryResult result = floodFill(x, y, new boolean[rows][cols]);
        return result.points;
    }

    Map<Player, Set<Point>> getTerritories() {
        Map<Player, Set<Point>> territories = new java.util.HashMap<>();
        territories.put(Player.BLACK, new java.util.HashSet<>());
        territories.put(Player.WHITE, new java.util.HashSet<>());
        territories.put(Player.NONE, new java.util.HashSet<>());

        boolean[][] globalVisited = new boolean[rows][cols];

        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < cols; x++) {
                if (!globalVisited[y][x] && grid[y][x] == ' ') {
                    boolean[][] localVisited = new boolean[rows][cols];
                    TerritoryResult result = floodFill(x, y, localVisited);
                    // mark all points as visited globally
                    for (Point p : result.points) {
                        globalVisited[p.y][p.x] = true;
                    }
                    territories.get(result.owner).addAll(result.points);
                }
            }
        }
        return territories;
    }

    private TerritoryResult floodFill(int startX, int startY, boolean[][] visited) {
        java.util.Set<Point> territory = new java.util.HashSet<>();
        java.util.Set<Character> borderingColors = new java.util.HashSet<>();
        java.util.Queue<Point> queue = new java.util.LinkedList<>();
        queue.add(new Point(startX, startY));

        while (!queue.isEmpty()) {
            Point p = queue.poll();
            int x = p.x;
            int y = p.y;

            if (x < 0 || y < 0 || y >= rows || x >= cols) continue;
            if (visited[y][x]) continue;
            visited[y][x] = true;

            char cell = grid[y][x];
            if (cell == ' ') {
                territory.add(p);
                queue.add(new Point(x + 1, y));
                queue.add(new Point(x - 1, y));
                queue.add(new Point(x, y + 1));
                queue.add(new Point(x, y - 1));
            } else if (cell == 'B' || cell == 'W') {
                borderingColors.add(cell);
            }
        }

        Player owner;
        if (borderingColors.size() == 1) {
            char c = borderingColors.iterator().next();
            owner = (c == 'B') ? Player.BLACK : Player.WHITE;
        } else {
            owner = Player.NONE;
        }

        return new TerritoryResult(owner, territory);
    }

    private static class TerritoryResult {
        final Player owner;
        final Set<Point> points;

        TerritoryResult(Player owner, Set<Point> points) {
            this.owner = owner;
            this.points = points;
        }
    }
}